api/proto/teleport/workloadidentity/v1/revocation_service.proto

```diff
@@ -0,0 +1,112 @@
+  // GetWorkloadIdentityX509Revocation retrieves a X509 workload identity
+  // revocation by name.
+  rpc GetWorkloadIdentityX509Revocation(GetWorkloadIdentityX509RevocationRequest) returns (WorkloadIdentityX509Revocation);
+  // DeleteWorkloadIdentityX509Revocation deletes a X509 workload identity
+  // revocation by name.
+  rpc DeleteWorkloadIdentityX509Revocation(DeleteWorkloadIdentityX509RevocationRequest) returns (google.protobuf.Empty);
+  // ListWorkloadIdentityX509Revocations returns a list of X509 workload
+  // identity revocations, pagination semantics are applied.
+  rpc ListWorkloadIdentityX509Revocations(ListWorkloadIdentityX509RevocationsRequest) returns (ListWorkloadIdentityX509RevocationsResponse);
+  // StreamSignedCRL returns the latest signed CRL to the client, and, any
+  // later updates to this signed CRL.
+  rpc StreamSignedCRL(StreamSignedCRLRequest) returns (stream StreamSignedCRLResponse);
+}
+
+// The request for CreateWorkloadIdentityX509Revocation.
+message CreateWorkloadIdentityX509RevocationRequest {
+  // The X509 workload identity revocation to create.
+  WorkloadIdentityX509Revocation workload_identity_x509_revocation = 1;
+}
+
+// The request for UpdateWorkloadIdentityX509Revocation.
+message UpdateWorkloadIdentityX509RevocationRequest {
+  // The X509 workload identity revocation to update.
+  WorkloadIdentityX509Revocation workload_identity_x509_revocation = 1;
+}
+
+// The request for UpsertWorkloadIdentityX509Revocation.
+message UpsertWorkloadIdentityX509RevocationRequest {
+  // The X509 workload identity revocation to upsert.
+  WorkloadIdentityX509Revocation workload_identity_x509_revocation = 1;
+}
+
+// The request for GetWorkloadIdentityX509Revocation.
+message GetWorkloadIdentityX509RevocationRequest {
+  // The name of the X509 workload identity revocation to retrieve.
+  string name = 1;
+}
+
+// The request for DeleteWorkloadIdentityX509Revocation.
+message DeleteWorkloadIdentityX509RevocationRequest {
+  // The name of the workload identity to delete.
+  string name = 1;
+}
+
+// The request for ListWorkloadIdentityX509Revocations.
+message ListWorkloadIdentityX509RevocationsRequest {
+  // The maximum number of items to return.
+  // The server may impose a different page size at its discretion.
+  int32 page_size = 1;
+  // The page_token value returned from a previous ListWorkloadIdentities request, if any.
+  string page_token = 2;
+}
+
+// The response for ListWorkloadIdentityX509Revocations.
+message ListWorkloadIdentityX509RevocationsResponse {
+  // The page of workload identities that matched the request.
+  repeated WorkloadIdentityX509Revocation workload_identity_x509_revocations = 1;
+  // Token to retrieve the next page of results, or empty if there are no
+  // more results in the list.
+  string next_page_token = 2;
+}
+
+// The request for StreamSignedCRL.
+message StreamSignedCRLRequest {}
```

Comment:
Based on the signature/messages I would assume it always starts with the complete CRL, saves a timestamp and then streams any changes as another complete CRL? I think it's worth spelling some of this out in the docs, wherever you think best (RPC or requests).

How large is the CRL expected to get? Should we design the ability to add/remove entries instead of always streaming a complete copy?

Is the CRL smart enough to ignore revocations for certificates that would already have expired anyway?

Reply:
f3b114b92e

> How large is the CRL expected to get? Should we design the ability to add/remove entries instead of always streaming a complete copy?

Unfortunately, for now, we'll likely need to stick with sending a full copy. Whilst Delta CRLs are specified, validator support is pretty spotty. In my PoC, I added a debounce mechanism to the signer to "group" revocations made within a few seconds of another together in a single signing which accounts for the "I have an automation calling `tctl` with a list of serials" use-case which would probably be the most intense.